### ISSUE TITLE
Whisper transcription unreliable: greedy_decode lacks in-loop logit suppression; long-form + transcribe CLI missing

### ISSUE THREAD

**Issue by Pushkinist** [Pushkinist/rMLX#119]
## Summary

Whisper transcription (`POST /v1/audio/transcriptions`) is **unreliable on real speech**. On most audio windows it returns an **empty transcript**; on some it emits **wrong-script garbage** (e.g. Cyrillic tokens for English audio); only occasionally does it produce correct text. The model loads correctly and the mel front-end is fine — the defect is in the decode loop's logit handling. Compounding this, there is **no long-form support** (audio >30 s is silently truncated) and **no real-audio test coverage** (the referenced smoke test does not exist).

Found by dogfooding `rmlx 0.2.1` on a real meeting recording.

## Repro

```sh
# any container -> 16 kHz mono WAV (server requires 16 kHz mono, <=25 MiB)
ffmpeg -y -ss 0 -i meeting.m4a -t 30 -ac 1 -ar 16000 -c:a pcm_s16le clip30.wav

rmlx serve \
  --whisper-model-path  $RMLX_O_MODELS_ROOT/mlx-community__whisper-large-v3-mlx \
  --whisper-tokenizer-path $RMLX_O_MODELS_ROOT/openai__whisper-large-v3-tokenizer \
  --port 8123

curl -s -X POST http://127.0.0.1:8123/v1/audio/transcriptions \
  -F file=@clip30.wav -F model=w -F language=en -F response_format=text
```

## Evidence

Same English meeting source, sliced into windows. Model: `whisper-large-v3-mlx`. `language=en` (explicit; `auto` behaves identically). Output content omitted — only sizes/behaviour shown.

| window | chars out | decode_ms | result |
|---|---|---|---|
| 10 s (offset 5 s) | 63 | 489 | ✅ correct text (verified vs reference transcript) |
| 10 s (offset 0 s) | 1 | 248 | ❌ single punctuation char |
| 15 s (offset 5 s) | 9 | 247 | ❌ wrong-script (Cyrillic) tokens for English audio |
| 20 / 30 / 40 / 60 s | 0 | 39–224 | ❌ empty |

Healthy cases run at RTF ≈ 0.05 (`decode_ms` ≈ 489 for 10 s); the empty cases finish in ~40 ms (`decode_ms`), i.e. the decoder halts almost immediately. The same source transcribes cleanly end-to-end under the Python `mlx_whisper` reference, so this is an rMLX decode bug, not a model or input problem.

## Root cause

`crates/rmlx-audio/src/whisper.rs::greedy_decode`:

- **`suppress_eot_at_prefill` is applied only at the first text position** (line ~824). No logit suppression is applied on subsequent steps.
- **Any timestamp token (`>= TOK_TIMESTAMP_BEGIN`) is treated as a hard stop**: empty return at prefill (lines ~836–843), `break` mid-loop (lines ~846–850).
- The in-loop `sample_next` receives **raw logits** (line ~868) — no `SuppressTokens` (non-speech / special-token set), and in `no_timestamps` mode no suppression of timestamp tokens.

Consequence: for dense/continuous speech the decoder samples a timestamp token early → stop → **empty**; with nothing constraining the output space it sometimes samples special/foreign tokens → **garbage**. A short clip with trailing silence happens to flow text before any timestamp token, which is why it occasionally works.

Reference behaviour (openai-whisper / `mlx_whisper` `DecodingTask`): apply logit filters at **every** step — `SuppressBlank` (first step), `SuppressTokens` (the non-speech/special set), and in `no_timestamps` mode suppress all timestamp tokens; in timestamp mode apply `apply_timestamp_rules`. rMLX applies none of these in the loop.

## Additional defects found alongside

- **No long-form.** `mel.rs::pad_or_trim` trims to 30 s (480 000 samples); only the first 30 s of any input is ever transcribed. VTT/SRT output is hardcoded to a single `00:00.000 --> 00:30.000` block (`audio.rs`), so there are no real per-segment timestamps.
- **`alignment_heads` npy fails to parse** at load (`NPY parse (alignment_heads): cannot parse dtype`, `npz.rs`) → word-level timestamps unavailable.
- **`verbose_json` reports `duration: 0.0`** regardless of input length.

## Proposed fix (phased — separate PRs)

- **P0 — decode correctness (blocker).** Add the standard Whisper logit filters to the `greedy_decode` step loop: `SuppressBlank` (first step), `SuppressTokens` (non-speech/special set), and `no_timestamps`-mode timestamp suppression (or `apply_timestamp_rules` in timestamp mode). Stop treating timestamp tokens as an unconditional halt. Acceptance: a battery of 30 s windows from a real recording all transcribe correctly and deterministically (temp=0).
- **P1 — long-form.** Sliding 30 s windows (wire the existing `vad.rs`, or fixed stride + overlap) with previous-text conditioning and cumulative real timestamps; emit multi-segment VTT/SRT. Acceptance: full-length transcript, WER vs a reference transcript within target.
- **P2 — `rmlx transcribe` CLI.** A subcommand `rmlx transcribe <audio> --model <snapshot> [--format vtt|srt|json|txt] [--language ...]`, arch-dispatched on `config.json` (Whisper first, seam for future ASR — **not** hardcoded to Whisper), in-process, honoring the single-MLX claim. Route the HTTP endpoint through the same long-form core.
- **P3 — polish.** Fix `alignment_heads` dtype parsing, `verbose_json` duration reporting.

## Test-coverage gap (and an env-var to remove)

Real-audio transcription has **no test coverage**: `crates/rmlx-audio/src/whisper_tests.rs` references a smoke test in `crates/rmlx-audio/tests/smoke.rs`, but **that file does not exist**. The unit tests only cover config/npy parsing — which is why a fully broken decode path shipped in 0.2.1.

The existing convention also gates such tests behind a bespoke `RMLX_TEST_MODEL_WHISPER` env var (referenced in `whisper_tests.rs`, `tokenizer_tests.rs`, `docs/MODELS.md`). This is the kind of single-purpose env knob the project rules discourage, and because nobody sets it in CI the gate guarantees the test never runs. Replace it with the standard `RMLX_O_MODELS_ROOT` snapshot auto-discovery used by `make model-check-full` (resolve the Whisper snapshot from the models root, run if present, **skip gracefully** if absent), and drop `RMLX_TEST_MODEL_WHISPER`. Add a non-gated real-audio regression smoke as part of P0 so this cannot silently rebreak.
